index.js

```diff
@@ -0,0 +1,375 @@
+ */
+internals.loadRemote = (authConfig) => {
+    const mutex = semaphore(1);
+    let keys = {};
+
+    return {
+        /**
+         * Load the key data (from the remote server if needed)
+         * @method getKey
+         * @param  {String}   keyId Key you are looking for
+         * @param  {Function} next  Function to call when done (error, key)
+         */
+        getKey: (keyId, next) => {
+            const nextSem = (err, result) => {
+                mutex.leave();
+                next(err, result);
+            };
+
+            // Prevent multiple calls to the remote URI
+            mutex.take(() => {
+                if (keys[keyId]) {
+                    return nextSem(null, keys[keyId]);
+                }
+
+                return request({
+                    url: authConfig.remoteUri,
+                    headers: {
+                        'user-agent': `${pkg.name}@${pkg.version}`,
+                        'hapi-auth-requested-key': keyId
+                    },
+                    json: true,
+                    timeout: authConfig.timeout
+                }, (reqError, response, body) => {
+                    if (reqError || response.statusCode !== 200) {
+                        const prevError = reqError ||
+                            new Error(`Status Code: ${response.statusCode}`);
+
+                        return nextSem(prevError);
+                    }
+
+                    keys = {};
+                    body.forEach((key) => {
+                        keys[guid(key.public)] = key;
+                    });
+
+                    if (keys[keyId]) {
+                        return nextSem(null, keys[keyId]);
+                    }
+
+                    return nextSem(new Error('Specified Key Id not found'));
+                });
+            });
+        }
+    };
+};
+
+/**
+ * Configures the authentication plugin of Hapi
+ * @method implementation
+ * @param  {HapiServer} server           HapiServer we are attaching the auth service to
+ * @param  {Object}     config           Plugin configuration
+ * @param  {Array}      config.keys      List of private keys to sign JWTs with
+ * @param  {String}     config.maxAge    Maximum age for signing JWTs
+ * @param  {String}     config.remoteUri URI of remote keys
+ * @param  {String}     config.timeout   How long to wait for API response
+ * @param  {String}     config.issuer    Name of issuing server
+ * @return {Object}                      Contains an authenticate function to parse each request
+ */
+internals.implementation = (server, config) => {
+    const authConfig = Joi.attempt(config, internals.SCHEMA_KEYS,
+        'Invalid config for hapi-auth-dynamic-jwt plugin');
+    let authObject = {};
+
+    // Local mode
+    if (typeof authConfig.keys === 'object') {
+        authObject = internals.loadLocal(authConfig);
+    } else {
+        authObject = internals.loadRemote(authConfig);
+    }
+
+    return {
+        api: authObject.api,
```

Comment:
I'm not sure how this works. So returning `api` here will automatically allow us to call the functions defined in api (`createJWT` and `availableKeys`)?

Also this isn't available for Remote. What happens when you load Remote Config then call `createJWT()`? Would it know to throw an error?

Reply:
Correct.  For remote mode those functions won't exist.